library/src/scala/internal/quoted/Matcher.scala

```diff
@@ -277,6 +301,31 @@ private[quoted] object Matcher {
       }
     end treeOps
 
+    private object ClosedTerm {
+      def unapply(term: Term)(given Context, Env): Option[term.type] =
+        if freeVars(term).isEmpty then Some(term) else None
+
+      def freeVars(tree: Tree)(given qctx: Context, env: Env): Set[Symbol] =
+        val accumulator = new TreeAccumulator[Set[Symbol]] {
+          def foldTree(x: Set[Symbol], tree: Tree)(given ctx: Context): Set[Symbol] =
+            tree match
+              case tree: Ident if env.contains(tree.symbol) => foldOverTree(x + tree.symbol, tree)
+              case _ => foldOverTree(x, tree)
+        }
+        accumulator.foldTree(Set.empty, tree)
+    }
```

Comment:
I'm not sure if `ClosedTerm` is the correct name, as it may still contain free variables. It's only free relative to `env`.